File: debuglogger/src/main/java/io/github/japskiddin/debuglogger/LogManager.java
package io.github.japskiddin.debuglogger;

import java.util.ArrayList;
import java.util.List;

public class LogManager {
  private static LogManager instance;
  private boolean enabled;
  private final List<LogEvent> logs;

  public static LogManager getInstance() {
    LogManager localInstance = instance;
    if (localInstance == null) {
      synchronized (LogManager.class) {
        localInstance = instance;
        if (localInstance == null) {
          instance = localInstance = new LogManager();
        }
      }
    }
    return localInstance;
  }

  public static void init() {
    if (instance == null) {
      instance = new LogManager();
    }
  }

  private LogManager() {
    logs = new ArrayList<>();
    enabled = false;
  }

  public List<LogEvent> getLogs() {
    return logs;
  }

  public void clear() {
    logs.clear();
  }

  public void addToLog(String tag, String event) {
    if (!enabled) return;
    LogEvent log = new LogEvent(tag, event);
    logs.add(log);
  }

  public boolean isEnabled() {
    return enabled;
  }

  public void setEnabled(boolean enabled) {
    this.enabled = enabled;
  }
}

File: debuglogger/src/main/java/io/github/japskiddin/debuglogger/LogManager.kt
package io.github.japskiddin.debuglogger

class LogManager private constructor() {
  var isEnabled = false
  private val logs: MutableList<LogEvent>

  init {
    logs = ArrayList()
  }

  fun getLogs(): List<LogEvent> {
    return logs
  }

  fun clear() {
    synchronized(logs) { logs.clear() }
  }

  fun logInfo(tag: String, event: String) {
    addToLog(LogEvent.LOG_INFO, tag, event)
  }

  fun logError(tag: String, event: String) {
    addToLog(LogEvent.LOG_ERROR, tag, event)
  }

  fun logDebug(tag: String, event: String) {
    addToLog(LogEvent.LOG_DEBUG, tag, event)
  }

  fun logWarn(tag: String, event: String) {
    addToLog(LogEvent.LOG_WARN, tag, event)
  }

  private fun addToLog(type: Int, tag: String, event: String) {
    if (!isEnabled) return
    synchronized(logs) {
      val log = LogEvent(type, tag, event)
      logs.add(log)
    }
  }

  companion object {
    private var instance: LogManager? = null
    @JvmStatic fun getInstance(): LogManager? {
      var localInstance = instance
      if (localInstance == null) {
        synchronized(LogManager::class.java) {
          localInstance = instance
          if (localInstance == null) {
            localInstance = LogManager()
            instance = localInstance
          }
        }
      }
      return localInstance
    }

    @JvmStatic fun init() {
      if (instance == null) {
        instance = LogManager()
      }
    }
  }
}